### PR TITLE
[WIP] fix: send hook yaml into post-renderer (#7891)

### LINKED ISSUE #7891: Post-renderer not receiving chart hook manifests?
Output of `helm version`: 

```
version.BuildInfo{Version:"v3.1.2", GitCommit:"d878d4d45863e42fd5cff6743294a11d28a9abce", GitTreeState:"clean", GoVersion:"go1.13.8"}
```

Output of `kubectl version`:

```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4+k3s1", GitCommit:"3eee8ac3a1cf0a216c8a660571329d4bda3bdf77", GitTreeState:"clean", BuildDate:"2020-03-25T16:13:25Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.4+k3s1", GitCommit:"3eee8ac3a1cf0a216c8a660571329d4bda3bdf77", GitTreeState:"clean", BuildDate:"2020-03-25T16:13:25Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"linux/amd64"}
```

I was experimenting with the `--post-renderer` option and a simple chart that includes one pod:

```
apiVersion: v1
kind: Pod
metadata:
  name: post-render-test
  labels:
    # The "app.kubernetes.io/managed-by" label is used to track which tool
    # deployed a given chart. It is useful for admins who want to see what
    # releases a particular tool is responsible for.
    app.kubernetes.io/managed-by: {{.Release.Service | quote }}
    # The "app.kubernetes.io/instance" convention makes it easy to tie a release
    # to all of the Kubernetes resources that were created as part of that
    # release.
    app.kubernetes.io/instance: {{.Release.Name | quote }}
    app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
    # This makes it easy to audit chart usage.
    helm.sh/chart: "{{.Chart.Name}}-{{.Chart.Version}}"
    values: {{.Values.Name}}
spec:
  # This shows how to use a simple value. This will look for a passed-in value
  # called restartPolicy. If it is not found, it will use the default value.
  # {{default "Never" .restartPolicy}} is a slightly optimized version of the
  # more conventional syntax: {{.restartPolicy | default "Never"}}
  restartPolicy: {{default "Never" .Values.restartPolicy}}
  containers:
  - name: waiter
    image: "alpine:3.4"
    command: ["/bin/sleep","9000"]
```

and one pre-install hook:

```
apiVersion: batch/v1
kind: Job
metadata:
  name: "{{ .Release.Name }}"
  labels:
    app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
    app.kubernetes.io/instance: {{ .Release.Name | quote }}
    app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
    helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
  annotations:
    # This is what defines this resource as a hook. Without this line, the
    # job is considered part of the release.
    "helm.sh/hook": pre-install,pre-upgrade
    "helm.sh/hook-weight": "-5"
    "helm.sh/hook-delete-policy": before-hook-creation
spec:
  backoffLimit: 3 # retry this task 3 times
  template:
    metadata:
      name: "{{ .Release.Name }}"
      labels:
        app.kubernetes.io/managed-by: {{ .Release.Service | quote }}
        app.kubernetes.io/instance: {{ .Release.Name | quote }}
        helm.sh/chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
    spec:
      restartPolicy: Never
      containers:
      - name: pre-deploy-job
        image: "alpine:3.4"
{{- if .Values.predeploy_task.fail }}
        command: ["/bin/false"]
{{- else }}
        command: ["/bin/sleep","{{ default "10" .Values.sleepyTime }}"]
{{- end }}
```

For the test, I was using a no-op post-renderer:

```
#!/bin/bash

tee --append  <&0 all.yaml
```

When I execute `helm install post-render-test ../chart --post-renderer ./post-renderer --debug --dry-run` I see both the chart hook and the pod manifest in the output, however my post-renderer only captures the pod manifest.

I've tried this as well with `helm template` with the same results.

Is this the intended behavior?  I couldn't find anything in the documentation to suggest that hooks would not be included when invoking the post-renderer.

### PR DESCRIPTION
**Pull request by kabakaev**
Fixes #7891.

Todo:
- [x] Rewrite `runPostRenderer()` to preserve YAML comments.
- [ ] Update unit tests.

### CODE REVIEW

**Comment by kabakaev** [comment]
@technosophos, @thomastaylor312, could you please take a look. 
I tried to implement [your suggestion from the corresponding issue](https://github.com/helm/helm/issues/7891#issuecomment-611780768).

In case you read this via email, i've also asked some questions [in the PR](https://github.com/helm/helm/pull/7948#issue-405519416).

File: pkg/action/install.go
Comment (thomastaylor312) [inline pkg/action/install.go:655]: This function name is a bit misleading as it both sets and removes annotations. I would split this functionality into 2 separate functions

File: pkg/action/install.go
Comment (thomastaylor312) [inline pkg/action/install.go:607]: This should be a `const` and not defined in the function. It should probably also be prefaced with a `helm.sh/`

File: pkg/action/install.go
Comment (thomastaylor312) [inline pkg/action/install.go:619]: This is an interesting approach. My one concern about this is what happens if the post-render step mucks with the annotations (either removing unknown annotations or setting their own). Are there any other ways we could do this without an annotation?

File: pkg/action/install.go
Comment (thomastaylor312) [inline pkg/action/install.go:635]: Should this be using the regex from above to split?

File: pkg/action/install.go
Comment (thomastaylor312) [inline pkg/action/install.go:666]: If there is no metadata for some weird reason, we shouldn't need to fail here. Instead I'd initialize an empty `map[string]interface{}` for the metadata

File: pkg/action/install.go
Comment (thomastaylor312) [inline pkg/action/install.go:684]: When splitting this out to its own function for removing, you probably don't need to return an error here. I don't think it indicates any real problem if there is no annotation to remove

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Thanks for working on this! There are a few changes requested here. Also, if you could add a simple post render test that makes sure it gets hooks from the chart, that would be great

**Comment by thomastaylor312** [comment]
Also, both of these points:
> func (c *Configuration) postProcessFiles( or func postProcessFiles(? Since this function is used only once, it's not really a separate piece of code. It just makes the main function more readable (smaller). Thus the c may be used in future refactoring, such as with c.Log(). As of now, c is not used.

> Variable and function names are quite verbose. Should they be shortened? Suggestions are welcome :)

I think everything is fine as is.

The documentation will need to be updated when this is merged. It can be found at: https://github.com/helm/helm-www/blob/master/content/en/docs/topics/advanced.md

File: pkg/action/install.go
Comment (kabakaev) [reply pkg/action/install.go:619]: The options are:
1. Run the post-renderer separately for non-hook templates (`[]byte`) and for each hook.

2. Remember template filename string for each object that is being marshalled for post-processing. For example, before postprocessing, fill the `map[string]string`, where `key:=fmt.Sprintf("%s;%s;$s", apiVersion, kind, metaName)` and value is the filename. Upon unmarshalling, use that map to find the filename. If not found, assign some random string.

3. Use [any other field from metadata](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#objectmeta-v1-meta), such as name or label. Unfortunately, other fields have more strict limitations on length and a set of allowed characters. IMHO, `annotations` is the best option there.

4. Use annotations, but if our `helm.sh/separator` annotation is not found, then "adopt" such object with a random file name. Are the original template filenames important after all? Don't know how they usedapart from a comment in the final yaml.

5. Mix options 2 and 4 by running post-renderer twice. Use first run with extra annotation to obtain extra mapping. In second run, post-render without extra annotations and assign a filename which best matches both mappings.

Option 4 is implemented now.

Option 5 may be dangerous if a user's postrenderer script is not idempotent.

I personally like the second option too. `kind` almost always stays the same. `apiVersion` often gets changed to a similar value. `metadata.name` is usually prefixed or suffixed with a string. So it may worth a try. I just don't know (yet) how to implemement a string similarity algorithm which works good for `metadata.name` and especially for `apiVersion` comparison.

File: pkg/action/install.go
Comment (kabakaev) [reply pkg/action/install.go:635]: I though initially that these document separators are guaranteed to be `"---\n"`, because they are added in the same function a few lines above. So I thought `strings.Split()` would save a few CPU cycles compared to `regexp.Split()`.
But now when you asked, it is indeed possible that post-renderer adds own separators that may have whitespaces, which may indent (and invalidate) the line after separator in case a simple `strings.Split()` is used. You're right, `regexp.Split()` must be used there.

File: pkg/action/install.go
Comment (kabakaev) [reply pkg/action/install.go:684]: I already shoot myself in the foot because of that error, when Kustomize post-renderer added a new networkPolicy object  :)

**Comment by kabakaev** [comment]
@thomastaylor312, PTAL! The post-renderer PR is updated according to your suggestions.

The test was also added:
```bash
# git checkout hook-post-renderer
# go test ./pkg/action/...
ok      helm.sh/helm/v3/pkg/action      0.090s
## revert changes to current master
# wget https://github.com/helm/helm/raw/master/pkg/action/action.go -O ./pkg/action/action.go
# go test ./pkg/action/...
---
# Source: hello/templates/hello
hello: world

--- FAIL: TestInstallRelease_PostRendererReceivesHooks (0.00s)
    install_test.go:319: Failed install: error while running post render on files: Hook was not post-processed
```

It seems the temporary annotation way works fine for a couple of "beta-testers" on the original issue #7891. But if you find it useful to add the failback mechanism described above in option `2. Remember template filename string...`, then I'd suggest to add it as a separate PR. Otherwise let's postpone it till someone claims a real-world usage scenario which gets broken by the temp annotation way.

File: pkg/action/action.go
Comment (thomastaylor312) [inline pkg/action/action.go:327]: Hopefully this never happens, but I wouldn't error here. If the yaml got so messed up that this doesn't have a metadata any more, we can let the kubernetes serialization catch it

File: pkg/action/action.go
Comment (thomastaylor312) [inline pkg/action/action.go:333]: Nit: This variable would be more descriptive for future readers if it was called `fileName`

File: pkg/action/action.go
Comment (thomastaylor312) [inline pkg/action/action.go:345]: This should probably use a hash of the data or a UID of some kind. It is unlikely, but you can still have a chance of a number in a random set repeating and causing a collision when trying to write out files later on

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Almost there! Just a few more small things, and then I'll need another maintainer to review due to the size

**Comment by mattfarina** [comment]
Since the bug is in the 3.3.0 milestone I moved the PR there as well.

**Comment by mattfarina** [comment]
@kabakaev Can you make the changes that @thomastaylor312 suggested?

**Comment by bacongobbler** [comment]
3.3.0-rc.1 is being cut tomorrow, so we're removing this from the milestone.

File: pkg/action/action.go
Comment (kabakaev) [reply pkg/action/action.go:345]: Good point. Let's return the document as is.

File: pkg/action/action.go
Comment (kabakaev) [reply pkg/action/action.go]: sha1 is used instead.

**Review by thomastaylor312 — APPROVED** [review]
This is looking good to go! Thanks for addressing all the changes. This will need 1 more maintainer review before it can be merged

**Review by technosophos — CHANGES_REQUESTED** [review]
It appears that the method proposed here strips all comments out of YAML before it is sent to the renderer. If a template author inserts something meaningful in the comments (like YTT templates) that are meant to be used by the post-render, it appears that these are lost before the post-render step. The ability to do this was one of the selling points of the post-renderer step.

If comments are stripped, this is a breaking change.

Any solution that can be merged in the Helm 3 cannot have breaking changes. So this PR would need to be refactored to avoid that behavior.

**Comment by kabakaev** [comment]
@technosophos, thanks for review. 

I was not aware `ytt` exists let alone [it uses YAML comments for templating](https://get-ytt.io/#example:example-k8s-helm-ish).

I'll try to refactor the PR accordingly.

**Comment by technosophos** [comment]
If you don't mind delaying for a few days, @kabakaev , I think that the real underlying problem is that the `renderResources` function is just doing too many things. I would like to look to see if there is a sensible way to refactor it that would make it much easier for changes like yours.

**Comment by kabakaev** [comment]
That sounds great. It's interesting to see what would be a refactored `renderResources`.

**Comment by technosophos** [comment]
A quick update: I am about 50% of the way through a refactor of renderResources.

**Comment by will-beta** [comment]
any progress?

**Comment by bhiravabhatla** [comment]
Any update on this?

**Comment by ron1** [comment]
Still waiting for an update on this.

**Comment by renier** [comment]
@technosophos Is this held up waiting on [the renderResources refactor](https://github.com/helm/helm/pull/7948#issuecomment-664608458)? or is it waiting on @kabakaev to ensure [comments are not stripped](https://github.com/helm/helm/pull/7948#issuecomment-664606643).

**Comment by technosophos** [comment]
I have given up on the render refactor between now and Helm 4 because the underlying code keeps changing too quickly. But we definitely cannot merge the PR if it strips comments.

**Comment by master4044** [comment]
@technosophos @kabakaev What is the plan for this change? @kabakaev are you still working on finding a way to not strip the comments?

I'm working with helm and Kustomize where I'm attempting to insert a value into jobs created by helm. I'm doing this to enforce the cleanup of those jobs if the creator of the job didn't set a ttlSecondsAfterComplete themselves. Without the post-render hook working, helm simply ignores the modification to the yaml.

**Comment by ThomasSteinbach** [comment]
So disappointing to me. I rebuilt my complete workflow using the Helm post-renderer just to recognize 10% of the K8s objects are not processed - and then seing this potential solution on the horizon like a rainbow. The closer you come, the rainbow keeps its distance 😞

**Comment by Chili-Man** [comment]
@technosophos @kabakaev anything I can do to help get this change through the finish line?

**Comment by Fantaztig** [comment]
Also interested in this, are the missing unit tests the only thing missing atm?

**Comment by kuskoman** [comment]
I also need this change in helm, is there anything I can do to speed up merging this PR?

**Comment by gaborho** [comment]
We are hindered by this limitation as well. Would be great if hook manifests would be passed over to post renderer. Would appreciate if this PR would be soon merged.

**Comment by MatthewCane** [comment]
This limitation is still causing issues, is there likely to be any progress on this?

**Comment by michelesr** [comment]
I'm not sure this work is complete:

```
$ helm version                                                                 
version.BuildInfo{Version:"v3.7+unreleased", GitCommit:"726556c6ad9dc774c5a5d1bdfe513017f2d727c1", GitTreeState:"clean", GoVersion:"go1.19.2"}

$ helm template --debug chart --post-renderer cat                                  
install.go:178: [debug] Original chart version: ""
install.go:199: [debug] CHART PATH: /home/michele/Projects/foo/chart

panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
helm.sh/helm/v3/pkg/action.runPostRenderer(0xc00056b658?, {0x2930d60, 0xc00060a880})
        helm.sh/helm/v3/pkg/action/action.go:278 +0x60c
helm.sh/helm/v3/pkg/action.(*Configuration).renderResources(0xc000318880, 0xc0000e3900, 0x7?, {0x2657651, 0xc}, {0x0, 0x0}, 0x0, 0x0, 0x0, ...)
        helm.sh/helm/v3/pkg/action/action.go:168 +0x5e5
helm.sh/helm/v3/pkg/action.(*Install).RunWithContext(0xc00033c580, {0x29487e0?, 0xc000485800}, 0x2?, 0x1?)
        helm.sh/helm/v3/pkg/action/install.go:259 +0x77d
main.runInstall({0xc000319bc0?, 0x1, 0x4}, 0xc00033c580, 0xc00068fa80?, {0x2931ba0?, 0xc000014018})
        helm.sh/helm/v3/cmd/helm/install.go:265 +0x8d9
main.newTemplateCmd.func2(0xc000004500?, {0xc000319bc0?, 0x4?, 0x4?})
        helm.sh/helm/v3/cmd/helm/template.go:82 +0x22a
github.com/spf13/cobra.(*Command).execute(0xc000004500, {0xc000319b80, 0x4, 0x4})
        github.com/spf13/cobra@v1.2.1/command.go:856 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000673900)
        github.com/spf13/cobra@v1.2.1/command.go:974 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.2.1/command.go:902
main.main()
        helm.sh/helm/v3/cmd/helm/helm.go:87 +0x235
```

You can use the chart at this repo to reproduce: https://github.com/michelesr/skaffold-test-repo

**Comment by joejulian** [comment]
Doesn't this completely change the api for post-renderer? Prior to this change, post-renderer could be used without it providing any output. If I'm reading this right, this change will require the post-renderer to output yaml which helm will then reinterpret. Am I reading this wrong?

**Comment by fuog** [comment]
I am also interested and would be happy if the change is adopted. Thanks a lot for all who have contributed here

**Comment by kabakaev** [comment]
Closing ins favor of https://github.com/helm/helm/issues/7891#issuecomment-1507430101

**Pull request closed without merging**
